### DISCUSSION TITLE
[Question] ModuleNotFoundError: No module named 'omni.client'

### DISCUSSION THREAD

**Discussion by nowionlyseedaylight** [isaac-sim/IsaacLab#1107] [Q&A]
### Question
I was running the code from orbit_tasks. However, I keep getting the error message that the module 'omni.client' cannot be imported. How can I solve this problem? I thought that I installed all packages provided by orbit.

vision@vision-Z690-AORUS-PRO-DDR4:~/orbit$ ./orbit.sh -p source/extensions/omni.isaac.orbit_tasks/omni/isaac/orbit_tasks/manipulation/reach/reach_env_cfg.py

[INFO] Using python from: /home/vision/.local/share/ov/pkg/isaac_sim-2023.1.0-hotfix.1/python.sh
Traceback (most recent call last):
  File "/home/vision/orbit/source/extensions/omni.isaac.orbit_tasks/omni/isaac/orbit_tasks/manipulation/reach/reach_env_cfg.py", line 10, in <module>
    import omni.isaac.orbit.sim as sim_utils
  File "/home/vision/orbit/source/extensions/omni.isaac.orbit/omni/isaac/orbit/sim/__init__.py", line 29, in <module>
    from .converters import *  # noqa: F401, F403
  File "/home/vision/orbit/source/extensions/omni.isaac.orbit/omni/isaac/orbit/sim/converters/__init__.py", line 20, in <module>
    from .asset_converter_base import AssetConverterBase
  File "/home/vision/orbit/source/extensions/omni.isaac.orbit/omni/isaac/orbit/sim/converters/asset_converter_base.py", line 17, in <module>
    from omni.isaac.orbit.utils.assets import check_file_path
  File "/home/vision/orbit/source/extensions/omni.isaac.orbit/omni/isaac/orbit/utils/assets.py", line 23, in <module>
    import omni.client
ModuleNotFoundError: No module named 'omni.client'
There was an error running python

**Comment by Mayankm96** [comment]
The files in Python modules are not directly runnable since they depend on Omniverse to be running first. This is because Omniverse extensions are hot-reloaded when the application is launched. They aren't available directly as Python modules.

There are ideas on how to make this possible but right now they are in the works (which may or may not become a thing). To verify your configs, you will first need to launch the simulation application.

Please check here for more info: https://isaac-orbit.github.io/orbit/source/setup/developer.html#standalone-applications

**Reply by qpwo** [reply]
I think new link is https://isaac-sim.github.io/IsaacLab/main/source/tutorials/00_sim/launch_app.html

**Comment by nowionlyseedaylight** [comment]
@Mayankm96  Thanks for your response.
However, I'm pretty sure it's runnable in python modules (through ./orbit.sh -p ....), since it worked with the last version of orbit.
My colleague made the additional folder in the _orbit/source/extensions/omni.isaac.orbit_tasks/omni/isaac/orbit_tasks/manipulation_ to implement the customized code. And it was possible to run the code file via local terminal.

But now, with this new version (or I'm not sure whether it's the problem of my situation) I continually get the error message like above. Is there any possibility for the reason of that error to be the orbit installation process? (for example, the module like omni.client may have been omitted)

If not, could it be implemented through standalone applications?

**Comment by pengxiangzang** [comment]
I'm also confused about this problem. How did you solve it? Could you please give me some advice?

**Comment by Mayankm96** [comment]
Are you able to run the demos mentioned here?

> https://isaac-sim.github.io/IsaacLab/source/setup/sample.html#showroom

As mentioned in the docs, you can't execute module-level code since it expects hot-reloading of Omniverse extensions. This only happens after you launch the simulation app.

To execute a standalone script, it should do the following:

```python
"""Launch Isaac Sim Simulator first."""

import argparse

from omni.isaac.lab.app import AppLauncher

# add argparse arguments
parser = argparse.ArgumentParser(description="My dummy script.")
# append AppLauncher cli args
AppLauncher.add_app_launcher_args(parser)
# parse the arguments
args_cli = parser.parse_args()

# launch omniverse app
app_launcher = AppLauncher(args_cli)
simulation_app = app_launcher.app

"""Rest everything follows."""

# INSERT YOUR PYTHON CODE
```

**Reply by meghbhalerao** [reply]
I feel like this point (and I think this is a very important one), while mentioned in many different places in the documentation, is not made very clear. 
For example it may be useful (for example) to have a section called - "Running Standalone Applications" under which this point should be made extremely clear.
For example say, I want to make a custom environement, as a developer I would like to quickly prototype my code by instantiating my env and testing the functions, and I would need to know that I need to run the simulation app everytime I want to test the code, but I am not sure if this is mentioned anywhere. I may be missing something, but these are my thoughts! Thanks again for the package!

**Comment by pengxiangzang** [comment]
I can run the demos in standalone , but when it comes to franka.py in omni.isaac.orbit_assets of extensions, this problem comes up. So the problem is that I didn't understand the structure of the codes.Thank you for your reply and sorry to take up your time.I'm just starting to learn how to use the isaac lab. Did you follow some video tutorials? The official documentation really confuses me.

**Comment by yuanbovin** [comment]
I find that this question may be due to the environment configuration of the Python library. First, check the omni client library is installed or not. Then, run the command `source /home/admins/.local/share/ov/pkg/isaac-sim-4.2.0/kit/setup_python_env.sh`. So you can import omni.client
![截图 2025-01-07 19-02-23](https://github.com/user-attachments/assets/0aaa8955-dc19-4a50-b753-a53ca0f16b77)

**Comment by iamzweiaa** [comment]
same problem here, have you found the solution?
and i run the Isaac lab in a docker

**Comment by Djctionary** [comment]
**Context**
Ubuntu 22.04.5 LTS, NVIDIA 4070 Ti Super, Isaac Sim 4.5, IsaacLab 2.1.

**Symptom**
Running a standalone terrain script failed with:

```
ModuleNotFoundError: No module named 'omni.client'
```

**What I observed**
This seems related to import order: some `isaaclab.*` modules (which depend on `omni.*`) weren’t available until the simulator (Kit) was actually launched.

**What worked for me**
Move the `isaaclab.sim`, `isaaclab.terrains`, and terrain config imports *after* creating the `AppLauncher(...).app`. Once I did that, the script ran and the terrain imported normally. This pattern also matches IsaacLab’s standalone examples.

**Minimal pattern (trimmed):**

```python
import argparse
from isaaclab.app import AppLauncher

def main():
    parser = argparse.ArgumentParser()
    AppLauncher.add_app_launcher_args(parser)
    args_cli = parser.parse_args()

    app_launcher = AppLauncher(args_cli)
    simulation_app = app_launcher.app  # launch first

    # import modules that depend on omni.* after app is running
    import isaaclab.sim as sim_utils
    from isaaclab.terrains import TerrainImporter, TerrainImporterCfg
    from isaaclab.terrains.config.rough import ROUGH_TERRAINS_CFG

    sim = sim_utils.SimulationContext(sim_utils.SimulationCfg(dt=0.01, device=getattr(args_cli, "device", "cpu")))
    sim.set_camera_view(eye=[5, 5, 5], target=[0, 0, 0])

    terrain_imp_cfg = TerrainImporterCfg(
        num_envs=1, env_spacing=3.0, prim_path="/World/ground",
        terrain_type="generator",
        terrain_generator=ROUGH_TERRAINS_CFG.replace(curriculum=False, color_scheme="none"),
        debug_vis=False, visual_material=None,
    )
    TerrainImporter(terrain_imp_cfg)

    sim.reset()
    while simulation_app.is_running():
        simulation_app.update()
    simulation_app.close()

if __name__ == "__main__":
    main()
